### PR TITLE
UMA02 - Filled Relay

### PR DESCRIPTION
**Pull request by RodrigoPina407**
# uma-2

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this bot please describe it here.

### CODE REVIEW

File: UMA-Bots/Filled-Relay/README.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+(https://goerli.etherscan.io/tx/0xdf8eddcff5e3b724577ea0c404e221a9b7ad89403fcde92206b2f3a79a50eaba): (expect 2 findings: 2 `FilledRelay` events)
+
+- [0x97ab0e9eedf62c2d76fe591b15fe9e36f865b81751a7aa0864393386fb64b2c6]
+(https://goerli.etherscan.io/tx/0x97ab0e9eedf62c2d76fe591b15fe9e36f865b81751a7aa0864393386fb64b2c6): (expect 1 finding: 1 `FilledRelay` event and 0 `SetDepositQuoteTimeBuffer` event)
+
+### Ethereum
+The bot behaviour can be verified with the following transactions on Ethereum mainnet ( Contract address is:[0x4D9079Bb4165aeb4084c526a32695dCfd2F77381](https://etherscan.io/address/0x4D9079Bb4165aeb4084c526a32695dCfd2F77381)):
+
+- [0xeffb7de17fe9ba78a4ea11803b4a2fce964261d3ecb800b0cad1935acb51e400]
+(https://etherscan.io/tx/0xeffb7de17fe9ba78a4ea11803b4a2fce964261d3ecb800b0cad1935acb51e400): (expect 1 finding: 1 `FilledRelay` event)
```
Comment (0xtaf) [inline UMA-Bots/Filled-Relay/README.md:60]: this returns 2 findings

File: UMA-Bots/Filled-Relay/README.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+### Arbitrum
+
+The bot behaviour can be verified with the following transactions on Arbitrum mainnet ( Contract address is:[0xb88690461ddbab6f04dfad7df66b7725942feb9c](https://arbiscan.io/address/0xb88690461ddbab6f04dfad7df66b7725942feb9c)):
+
+- [0xfcdb0e4614849710f2195d2eb2f9b3c9fa424b3e12066288c10b39bfa514b2ac]
+(https://arbiscan.io/tx/0xfcdb0e4614849710f2195d2eb2f9b3c9fa424b3e12066288c10b39bfa514b2ac): (expect 1 finding: 1 `FilledRelay` event)
+
+- [0xfcdb0e4614849710f2195d2eb2f9b3c9fa424b3e12066288c10b39bfa514b2ac]
+(https://arbiscan.io/tx/0xfcdb0e4614849710f2195d2eb2f9b3c9fa424b3e12066288c10b39bfa514b2ac): (expect 2 findings: 2 `FilledRelay` events)
```
Comment (0xtaf) [inline UMA-Bots/Filled-Relay/README.md:76]: these are the same tx hashes

File: UMA-Bots/Filled-Relay/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import { FILLED_RELAY_EVENT } from "./ABI";
+import { NetworkManager } from "forta-agent-tools";
+import { NetworkData, DATA } from "./config";
+import { createFinding } from "./findings";
+import BN from "bignumber.js";
+import { getTokenInfo } from "./utils";
+import LRU from "lru-cache";
+
+const networkManager = new NetworkManager(DATA);
+let token_cache = new LRU<string, { tokenName: string; tokenDecimals: number }>({ max: 500 });
```
Comment (0xtaf) [inline UMA-Bots/Filled-Relay/src/agent.ts:11]:
```suggestion
const token_cache = new LRU<string, { tokenName: string; tokenDecimals: number }>({ max: 500 });
```

File: UMA-Bots/Filled-Relay/src/agent.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+      let metadata = {
+        amount: normalizedAmount.toString(10),
+        originChainId: originChainId.toString(),
+        destinationChainId: destinationChainId.toString(),
+        tokenName: tokenInfo.tokenName,
+        depositor,
+        recipient,
+        relayer,
+      };
```
Comment (0xtaf) [inline UMA-Bots/Filled-Relay/src/agent.ts:51]:
```suggestion
      const { amount, originChainId, destinationChainId, destinationToken, depositor, relayer, recipient, isSlowRelay } =
        filledRelayEvent.args;

      const tokenInfo: { tokenName: string; tokenDecimals: number } = await getTokenInfo(destinationToken, provider, token_cache, txEvent.blockNumber);

      const normalizedAmount = BN(amount.toString()).shiftedBy(-tokenInfo.tokenDecimals);

      const metadata = {
        amount: normalizedAmount.toString(10),
        originChainId: originChainId.toString(),
        destinationChainId: destinationChainId.toString(),
        tokenName: tokenInfo.tokenName,
        depositor,
        recipient,
        relayer,
      };
```

**Review by 0xtaf — CHANGES_REQUESTED** [review]
another great work 👍 the needed changes are minor and not logic related

File: UMA-Bots/Filled-Relay/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+      tokenName: "Test Token",
+      recipient: createAddress("0x0001"),
+      depositor: createAddress("0x0021"),
+      relayer: createAddress("0x6473"),
+    };
+
+    expect(findings).toStrictEqual([createFinding(metadata, false)]);
+  });
+
+  it("returns multiple finding if there are FilledRelay events emitted", async () => {
```
Comment (Vxatz) [inline UMA-Bots/Filled-Relay/src/agent.spec.ts:151]:
```suggestion
  it("returns multiple findings if there are FilledRelay events emitted", async () => {
```

**Review by Vxatz — CHANGES_REQUESTED** [review]
Besides a few minor comments, LGTM 👍

**Review by 0xtaf — APPROVED** [review]
👍

**Review by Vxatz — APPROVED** [review]

**Pull request merged**